### PR TITLE
veilid: init at 0.1.7

### PR DESCRIPTION
**Pull request by bbigras**
## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by evanrichter** [comment]
builds and runs for me on linux only right now. on aarch64-darwin, I get a seemingly unrelated build error in the cryptomator package. I have no idea how that is even relevant for this package though. (tagging the maintainer @bachp for any idea?)


1 package built:
veilid

Result of `nixpkgs-review pr 248581` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>veilid</li>
  </ul>
</details>

**Comment by evanrichter** [comment]
I would rather see versioning follow the tagged releases, seen [here](https://gitlab.com/veilid/veilid/-/tags). Otherwise, it's hard to know when to cut unstable releases

**Comment by qbit** [comment]
Result of `nixpkgs-review pr 248581` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>veilid</li>
  </ul>
</details>

**Comment by qbit** [comment]
```
error: builder for '/nix/store/barxxd0wbrpw9b18hbi9b7mizivx61iy-bugsalot-336a705.drv' failed with exit code 1;
       last 9 log lines:
       > exporting https://github.com/crioux/bugsalot.git (rev 336a7053faadf990b9362edf5752ef34fa1f9615) into /nix/store/4viihzaanlzcq8b1ywdpr07dan5g2qsr-bugsalot-336a705
       > Initialized empty Git repository in /nix/store/4viihzaanlzcq8b1ywdpr07dan5g2qsr-bugsalot-336a705/.git/
       > error: cannot run ssh: No such file or directory
       > fatal: unable to fork
       > error: cannot run ssh: No such file or directory
       > fatal: unable to fork
       > error: cannot run ssh: No such file or directory
       > fatal: unable to fork
       > Unable to checkout 336a7053faadf990b9362edf5752ef34fa1f9615 from https://github.com/crioux/bugsalot.git.
       For full logs, run 'nix log /nix/store/barxxd0wbrpw9b18hbi9b7mizivx61iy-bugsalot-336a705.drv'.

```

**Comment by evanrichter** [comment]
1 package failed to build:
veilid

Result of `nixpkgs-review pr 248581` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>veilid</li>
  </ul>
</details>

I updated my fork of nixpkgs and the veilid build went much farther! just a final linking step needed for darwin: `ld: framework not found Security`

**Comment by evanrichter** [comment]
with this patch, veilid builds and runs on aarch64-darwin:

```patch
diff --git a/pkgs/tools/networking/veilid/default.nix b/pkgs/tools/networking/veilid/default.nix
index bff75c5fc4d..b1614303734 100644
--- a/pkgs/tools/networking/veilid/default.nix
+++ b/pkgs/tools/networking/veilid/default.nix
@@ -1,4 +1,7 @@
 { lib
+, stdenv
+, AppKit
+, Security
 , fetchFromGitLab
 , rustPlatform
 , protobuf
@@ -29,6 +32,8 @@ rustPlatform.buildRustPackage rec {
     protobuf
   ];
 
+  buildInputs = lib.optionals stdenv.isDarwin [ AppKit Security ];
+
   cargoBuildFlags = [
     "--workspace"
   ];
diff --git a/pkgs/top-level/all-packages.nix b/pkgs/top-level/all-packages.nix
index ff5fbbebab5..3ee288cfa5f 100644
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -41009,7 +41009,9 @@ with pkgs;
 
   vazir-fonts = callPackage ../data/fonts/vazir-fonts { };
 
-  veilid = callPackage ../tools/networking/veilid { };
+  veilid = callPackage ../tools/networking/veilid {
+    inherit (darwin.apple_sdk.frameworks) AppKit Security;
+  };
 
   vhs = callPackage ../applications/misc/vhs { };
 
```

**Comment by wamserma** [comment]
Result of `nixpkgs-review pr 248581` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>veilid</li>
  </ul>
</details>

**Comment by bbigras** [comment]
I made all the changes.

Note that I also split the package into "out" and "lib", since lib is kinda big for nothing when you only want the bin.

@qbit I have no idea why your build failed. Could you try to update your nixpkgs in case it helps like in this case https://github.com/NixOS/nixpkgs/pull/248581#issuecomment-1676194585

**Comment by bbigras** [comment]
It seems like /nix/store/85hlq7pk8nzapgl7fc3wicgbdqwcw1sc-veilid-0.1.7 still depends on /nix/store/cxiwpf1y5jih9i0yqfxghhpxnvfwzi56-veilid-0.1.7-lib which is 565.5 MiB.

Anyone knows a way to nix copy only the bins?

**Comment by bbigras** [comment]
I wonder if it's a bad idea to do something like `moveToOutput "nix-support" "$dev"` or remove nix-support/propagated-build-inputs.

**Comment by qbit** [comment]
@bbigras Not sure either. Just tried from nixpkgs:d58f7137e0ed0dd90efebbea0254548bd0babcb6 - my system is at afde5d87aae5fe3438f120a44d88c5ab4c59abcb. 

The latest iteration still fails for me (veilid veilid.lib). Same error.

**Comment by qbit** [comment]
Building outside of `nixpkgs-review` works as expected.

File: pkgs/tools/networking/veilid/default.nix
Comment (AndersonTorres) [inline pkgs/tools/networking/veilid/default.nix:50]: Description should not be self-referent.

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/tools/networking/veilid/default.nix
Comment (evanrichter) [reply pkgs/tools/networking/veilid/default.nix:50]: from the project homepage:

> Veilid is an open-source, peer-to-peer, mobile-first, networked application framework.

so how about

> An open-source, peer-to-peer, mobile-first, networked application framework.

?

Comment (AndersonTorres) [reply pkgs/tools/networking/veilid/default.nix:50]: Fine.
Just remember to remove the `.` at the end.

Comment (evanrichter) [reply pkgs/tools/networking/veilid/default.nix:50]: Good to know!

Comment (bbigras) [reply pkgs/tools/networking/veilid/default.nix]: Done, thanks!

**Review by AndersonTorres — APPROVED** [review]

**Comment by AndersonTorres** [comment]
NOw it needs investigation - why is this failing on intel apple?

File: pkgs/tools/networking/veilid/default.nix
Comment (evanrichter) [inline pkgs/tools/networking/veilid/default.nix:7]: What is protobuf needed for? I made a dev flake for veilid and so far I've only needed capnproto

Comment (bbigras) [reply pkgs/tools/networking/veilid/default.nix:7]: the opentelemetry crate. Let me try without it.

Comment (bbigras) [reply pkgs/tools/networking/veilid/default.nix:7]: Here's when I remove it:

```
error: failed to run custom build command for `opentelemetry-proto v0.1.0`

Caused by:
  process didn't exit successfully: `/build/source/target/release/build/opentelemetry-proto-8628f340da5d6423/build-script-build` (exit status: 101)
  --- stdout
  cargo:rerun-if-changed=src/proto/opentelemetry-proto/opentelemetry/proto/common/v1/common.proto
  cargo:rerun-if-changed=src/proto/opentelemetry-proto/opentelemetry/proto/resource/v1/resource.proto
  cargo:rerun-if-changed=src/proto/opentelemetry-proto/opentelemetry/proto/trace/v1/trace.proto
  cargo:rerun-if-changed=src/proto/opentelemetry-proto/opentelemetry/proto/trace/v1/trace_config.proto
  cargo:rerun-if-changed=src/proto/opentelemetry-proto/opentelemetry/proto/collector/trace/v1/trace_service.proto
  cargo:rerun-if-changed=src/proto/opentelemetry-proto/opentelemetry/proto/metrics/v1/metrics.proto
  cargo:rerun-if-changed=src/proto/opentelemetry-proto/opentelemetry/proto/collector/metrics/v1/metrics_service.proto
  cargo:rerun-if-changed=src/proto/opentelemetry-proto/opentelemetry/proto/logs/v1/logs.proto
  cargo:rerun-if-changed=src/proto/opentelemetry-proto/opentelemetry/proto/collector/logs/v1/logs_service.proto
  cargo:rerun-if-changed=src/proto/opentelemetry-proto

  --- stderr
  thread 'main' panicked at 'Could not find `protoc` installation and this build crate cannot proceed without
      this knowledge. If `protoc` is installed and this crate had trouble finding
      it, you can set the `PROTOC` environment variable with the specific path to your
      installed `protoc` binary.If you're on debian, try `apt-get install protobuf-compiler` or download it from https://github.com/protocolbuffers/protobuf/releases

  For more information: https://docs.rs/prost-build/#sourcing-protoc
  ', /build/cargo-vendor-dir/prost-build-0.11.9/src/lib.rs:1457:10
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
warning: build failed, waiting for other jobs to finish...
error: builder for '/nix/store/0jzgnwd1fsvhkbjdphl44y25gy9v05k6-veilid-0.1.7.drv' failed with exit code 101;
       last 10 log lines:
       >   --- stderr
       >   thread 'main' panicked at 'Could not find `protoc` installation and this build crate cannot proceed without
       >       this knowledge. If `protoc` is installed and this crate had trouble finding
       >       it, you can set the `PROTOC` environment variable with the specific path to your
       >       installed `protoc` binary.If you're on debian, try `apt-get install protobuf-compiler` or download it from https://github.com/protocolbuffers/protobuf/releases
       >
       >   For more information: https://docs.rs/prost-build/#sourcing-protoc
       >   ', /build/cargo-vendor-dir/prost-build-0.11.9/src/lib.rs:1457:10
       >   note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
       > warning: build failed, waiting for other jobs to finish...
       For full logs, run 'nix log /nix/store/0jzgnwd1fsvhkbjdphl44y25gy9v05k6-veilid-0.1.7.drv'.
```

Comment (evanrichter) [reply pkgs/tools/networking/veilid/default.nix:7]: ah you're right! needed for veilid-server. I was just using veilid-core

**Comment by evanrichter** [comment]
> NOw it needs investigation - why is this failing on intel apple?

I don't know, but that symbol name seems to need a recent MacOS version. I'm not sure what ofborg is using..

#214872 seems relevant

**Comment by AndersonTorres** [comment]
Ping me 24 hours to merge

**Comment by bbigras** [comment]
I'm trying to remove "lib" as a dependency of "out"

```sh
# normal
❯ nix path-info --recursive --closure-size -h ./result
/nix/store/5cfwfnp4capzbjmkkrl7a6shvaghgj3h-xgcc-12.3.0-libgcc   139.1K
/nix/store/748caky83gr4pyrhircvi4llx99ja6z9-libunistring-1.1       1.8M
/nix/store/v5llm7rmipvg487yvgb15q0m2jxrihq3-libidn2-2.3.4          2.1M
/nix/store/ibp4camsx1mlllwzh32yyqcq2r2xsy1a-glibc-2.37-8          31.1M
/nix/store/kbjir9mdrz2w1f7kl4ikisg491x2xd1y-gcc-12.3.0-libgcc    139.1K
/nix/store/4igdc32rmnijcra8y3r1h42987ghzag2-gcc-12.3.0-lib        38.8M
/nix/store/cxiwpf1y5jih9i0yqfxghhpxnvfwzi56-veilid-0.1.7-lib     581.2M
/nix/store/85hlq7pk8nzapgl7fc3wicgbdqwcw1sc-veilid-0.1.7         604.3M

# with default "bin" output
❯ nix path-info --recursive --closure-size -h ./result-bin
/nix/store/5cfwfnp4capzbjmkkrl7a6shvaghgj3h-xgcc-12.3.0-libgcc   139.1K
/nix/store/748caky83gr4pyrhircvi4llx99ja6z9-libunistring-1.1       1.8M
/nix/store/v5llm7rmipvg487yvgb15q0m2jxrihq3-libidn2-2.3.4          2.1M
/nix/store/ibp4camsx1mlllwzh32yyqcq2r2xsy1a-glibc-2.37-8          31.1M
/nix/store/kbjir9mdrz2w1f7kl4ikisg491x2xd1y-gcc-12.3.0-libgcc    139.1K
/nix/store/4igdc32rmnijcra8y3r1h42987ghzag2-gcc-12.3.0-lib        38.8M
/nix/store/17j11f2i7pwvc7kpgmyz5jni5v19qyrk-veilid-0.1.7-bin      61.8M
```

**Comment by bbigras** [comment]
It seems adding a "dev" output fixes the size problem. Related to https://github.com/FRidh/nixpkgs/commit/7aacf5f93aec2b0bcc5cd9dac04bc8ea27ac49c3

```sh
❯ nix path-info --recursive --closure-size -h ./result
/nix/store/5cfwfnp4capzbjmkkrl7a6shvaghgj3h-xgcc-12.3.0-libgcc   139.1K
/nix/store/748caky83gr4pyrhircvi4llx99ja6z9-libunistring-1.1       1.8M
/nix/store/v5llm7rmipvg487yvgb15q0m2jxrihq3-libidn2-2.3.4          2.1M
/nix/store/ibp4camsx1mlllwzh32yyqcq2r2xsy1a-glibc-2.37-8          31.1M
/nix/store/kbjir9mdrz2w1f7kl4ikisg491x2xd1y-gcc-12.3.0-libgcc    139.1K
/nix/store/4igdc32rmnijcra8y3r1h42987ghzag2-gcc-12.3.0-lib        38.8M
/nix/store/7va4xbql210vrkza199wbxnsal673w44-veilid-0.1.7          61.8M
```

**Comment by evanrichter** [comment]
> I'm trying to remove "lib" as a dependency of "out"

Maybe change the build command from "--workspace" to "-p" "veilid-server" "-p" "veilid-cli"?

**Comment by bbigras** [comment]
> > I'm trying to remove "lib" as a dependency of "out"
> 
> Maybe change the build command from "--workspace" to "-p" "veilid-server" "-p" "veilid-cli"?

That would package lib, I still want it, Just not ship the 550MB with the bins.

The current version looks ok. -lib is still being built and closure size is 62MB

**Comment by bbigras** [comment]
@AndersonTorres ping

**Comment by bbigras** [comment]
Thanks everyone!

**Pull request merged**
